dist/rules/no-internal.js

```diff
@@ -82,21 +89,57 @@ module.exports = {
       return undefined;
     }
 
-    /** @param {string} fileName */
-    function isLocalFile(fileName) {
-      return fileName && typeof fileName === "string" && !fileName.includes("node_modules");
+    function dirContainsPath(dir, targetPath) {
+      const relative = path.relative(dir, targetPath);
+      return (
+        !!relative && !relative.startsWith("..") && !path.isAbsolute(relative)
+      );
+    }
+
+    function pathContainsCheckedPackage(packagePath) {
+      return checkedPackageRegexes.some((r) => r.test(packagePath));
+    }
+
+    function packageNameContainsCheckedPackage(packagePath) {
+      const packageList = workspace.getWorkspaces(packagePath);
+      // Look through all package infos to find the one containing our packagePath
+      for (let pkg in packageList) {
+        const packageObj = packageList[pkg];
+        const packageBaseDir = packageObj.packageJson.packageJsonPath.split("package.json")[0];
+
+        // Note: this returns false if packagePath == packageBaseDir
+        if (dirContainsPath(packageBaseDir, packagePath)) {
+          return pathContainsCheckedPackage(packageObj.name);
+        }
+      }
+
+      return false;
     }
 
+    // Returns true if a file is within a package for which @internal is a violation
+    // By default, @itwin and @bentley packages are included
     function isCheckedFile(declaration) {
       if (!declaration)
         return false;
       const fileName = getFileName(declaration.parent);
-      // eslint fileName always uses unix path separators
-      const packageSegments = fileName.split("node_modules/");
-      // can be undefined
-      const packagePath = packageSegments[packageSegments.length - 1];
-      const inCheckedPackage = packagePath && checkedPackageRegexes.some((r) => r.test(packagePath));
-      return inCheckedPackage && !isLocalFile(declaration);
+
+      if (fileName.includes('node_modules')) {
+        // If in node modules (real dependency), check path
+
+        // eslint fileName always uses unix path separators
+        const packageSegments = fileName.split("node_modules/");
+        // can be undefined
+        const packagePath = packageSegments[packageSegments.length - 1];
+        return packagePath && pathContainsCheckedPackage(packagePath);
+
+      } else if (allowWorkspaceInternal && !dirContainsPath(parserServices.program.getCommonSourceDirectory(), fileName)) {
```

Comment:
nit: imo would be clearer if you remove the `else` part of the `if` and instead did:
```ts
// above checked that the file isn't in an installed dependency
const isWorkspaceLinkedDependency = !dirContainsPath(parserServices.program.getCommonSourceDirectory(), fileName);

if (allowWorkspaceInternal && isWorkspaceLinkedDependency)
  return false;
```